### PR TITLE
K8sDeploymentManager status

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  private val progressingCondition = "Progressing"
+
+  //https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+  //"For some conditions, True represents normal operation, and for some conditions, False represents normal operation."...
+  //in our case Availability and Progressing have "positive polarity" as described in link above...
+  private val conditionStatusToCheck = "True"
+
+  private[manager] def findStatusForDeployments(deployments: List[Deployment]): Option[ProcessState] = {
+    def status(deployment: Deployment): ProcessState = {
+      val version = deployment.metadata.annotations.get(scenarioVersionAnnotation)
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala:24]: Maybe we can have both creation of this annotation and extraction in `K8sDeploymentManager` companion object to keep them synchronized?

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        .flatMap(CirceUtil.decodeJson[ProcessVersion](_).toOption)
+      val (status, attrs, errors) = deployment.status match {
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala:26]: wow I haven't known that we have `errors` next to `status`. pretty weird - interesting how they are presented for success statuses. I've added `EmbeddedStateStatus.DetailedFailedStateStatus` in some change - wondering if it was needed. Am i right that `attributes` are always `None`?

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 3]
```diff
@@ -106,6 +120,8 @@ class K8sDeploymentManager(modelData: ModelData,
             strategy = Some(Deployment.Strategy.Recreate),
             //here we use id to avoid sanitization problems
             selector = LabelSelector(IsEqualRequirement(scenarioIdLabel, processVersion.processId.value.toString)),
+            progressDeadlineSeconds = Some(20), //config.progressDeadlineSeconds,
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala]: :+1: true, I forgot to remove some debug stuff...

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  private val progressingCondition = "Progressing"
+
+  //https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+  //"For some conditions, True represents normal operation, and for some conditions, False represents normal operation."...
+  //in our case Availability and Progressing have "positive polarity" as described in link above...
+  private val conditionStatusToCheck = "True"
+
+  private[manager] def findStatusForDeployments(deployments: List[Deployment]): Option[ProcessState] = {
+    def status(deployment: Deployment): ProcessState = {
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: done

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+  //TODO: should we add responses to status attributes?
+  private[manager] def mapStatus(status: Deployment.Status): (StateStatus, Option[Json], List[String]) = {
+    def trueCondition(name: String): Option[(Deployment.Condition, Boolean)] = status.conditions
+      .find(cd => cd.`type` == name)
+      .map(k => (k, k.status == conditionStatusToCheck))
+
+    (trueCondition(availableCondition), trueCondition(progressingCondition)) match {
+      case (Some((_, true)), _) => (SimpleStateStatus.Running, None, Nil)
+      case (_, Some((_, true))) => (SimpleStateStatus.DuringDeploy, None, Nil)
+      case (a, b) => (SimpleStateStatus.Failed, None, a.flatMap(_._1.message).toList ++ b.flatMap(_._1.message).toList)
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: I refactored this piece a bit, have a look now. I still have a flatMap to handle some/none, none/some etc.

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        val errors = List(s"Expected one deployment, instead: ${duplicates.map(_.metadata.name).mkString(", ")}")
+        Some(ProcessState(None, K8sStateStatus.MultipleJobsRunning, None, definitionManager, None, None, errors))
+    }
+  }
+
+  //TODO: should we add responses to status attributes?
+  private[manager] def mapStatus(status: Deployment.Status): (StateStatus, Option[Json], List[String]) = {
+    def trueCondition(name: String): Option[(Deployment.Condition, Boolean)] = status.conditions
+      .find(cd => cd.`type` == name)
+      .map(k => (k, k.status == conditionStatusToCheck))
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: I refactored this piece a bit, have a look now.

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+class K8sDeploymentStatusMapper(definitionManager: ProcessStateDefinitionManager) {
+
+  private val availableCondition = "Available"
+
+  private val progressingCondition = "Progressing"
+
+  //https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties
+  //"For some conditions, True represents normal operation, and for some conditions, False represents normal operation."...
+  //in our case Availability and Progressing have "positive polarity" as described in link above...
+  private val conditionStatusToCheck = "True"
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: done.

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+import pl.touk.nussknacker.engine.api.{CirceUtil, ProcessVersion}
+import pl.touk.nussknacker.engine.api.deployment.{ProcessState, ProcessStateDefinitionManager, StateStatus}
+import pl.touk.nussknacker.engine.api.deployment.simple.SimpleStateStatus
+import pl.touk.nussknacker.k8s.manager.K8sDeploymentManager.scenarioVersionAnnotation
+import skuber.apps.v1.Deployment
+
+//Based on https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#deployment-status
+class K8sDeploymentStatusMapper(definitionManager: ProcessStateDefinitionManager) {
+
+  private val availableCondition = "Available"
```
[the remark this answers is not in the document]
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala:12]: done

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 1 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: done

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 2 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: Yes, attributes are always none. As for errors - I also often forget about them - they are always presented as popover I think, see ProcessStateIcon.tsx#imageWithPopover

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: looks simple enough, thanks :+1:

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: much more readable, thanks

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentStatusMapper.scala]: thanks

**Review by arkadius — APPROVED** [review]

**Pull request merged**